rclcpp/test/rclcpp/test_generic_client.cpp

```diff
@@ -0,0 +1,499 @@
+        node->get_node_graph_interface(),
+        node->get_node_services_interface(),
+        "test_service",
+        "test_msgs/srv/InvalidType",
+        rclcpp::ServicesQoS(),
+        nullptr);
+    }, std::runtime_error);
+  }
+}
+
+TEST_F(TestGenericClient, construct_with_rcl_error) {
+  {
+    // reset() is not necessary for this exception, but handles unused return value warning
+    auto mock = mocking_utils::patch_and_return("lib:rclcpp", rcl_client_init, RCL_RET_ERROR);
+    EXPECT_THROW(
+      node->create_generic_client("test_service", "test_msgs/srv/Empty").reset(),
+      rclcpp::exceptions::RCLError);
+  }
+  {
+    // reset() is required for this one
+    auto mock = mocking_utils::patch_and_return("lib:rclcpp", rcl_client_fini, RCL_RET_ERROR);
+    EXPECT_NO_THROW(
+      node->create_generic_client("test_service", "test_msgs/srv/Empty").reset());
+  }
+}
+
+TEST_F(TestGenericClient, wait_for_service) {
+  const std::string service_name = "test_service";
+  auto client = node->create_generic_client(service_name, "test_msgs/srv/Empty");
+  EXPECT_FALSE(client->wait_for_service(std::chrono::nanoseconds(0)));
+  EXPECT_FALSE(client->wait_for_service(std::chrono::milliseconds(10)));
+
+  auto callback = [](
+    const test_msgs::srv::Empty::Request::SharedPtr,
+    test_msgs::srv::Empty::Response::SharedPtr) {};
+
+  auto service =
+    node->create_service<test_msgs::srv::Empty>(service_name, std::move(callback));
+
+  EXPECT_TRUE(client->wait_for_service(std::chrono::nanoseconds(-1)));
+  EXPECT_TRUE(client->service_is_ready());
+}
+
+/*
+   Testing generic client construction and destruction for subnodes.
+ */
+TEST_F(TestGenericClientSub, construction_and_destruction) {
+  {
+    auto client = subnode->create_generic_client("test_service", "test_msgs/srv/Empty");
+    EXPECT_STREQ(client->get_service_name(), "/ns/test_service");
+  }
+
+  {
+    ASSERT_THROW(
+    {
+      auto client = node->create_generic_client("invalid_service?", "test_msgs/srv/Empty");
+    }, rclcpp::exceptions::InvalidServiceNameError);
+  }
+}
+
+class TestGenericClientWithServer : public ::testing::Test
+{
+protected:
+  static void SetUpTestCase()
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  static void TearDownTestCase()
+  {
+    rclcpp::shutdown();
+  }
+
+  void SetUp()
+  {
+    node = std::make_shared<rclcpp::Node>("test_node", "ns");
+
+    auto callback = [](
+      const test_msgs::srv::Empty::Request::SharedPtr,
+      test_msgs::srv::Empty::Response::SharedPtr) {};
+
+    service = node->create_service<test_msgs::srv::Empty>(service_name, std::move(callback));
+  }
+
+  ::testing::AssertionResult SendEmptyRequestAndWait(
+    std::chrono::milliseconds timeout = std::chrono::milliseconds(1000))
+  {
+    auto client = node->create_generic_client(service_name, "test_msgs/srv/Empty");
+    if (!client->wait_for_service()) {
+      return ::testing::AssertionFailure() << "Waiting for service failed";
```

Comment:
actually this does not wait, it just checks if service is online w/o waiting timeout.

```suggestion
      return ::testing::AssertionFailure() << "Service is not available yet";
```

Reply:
Yes. Updated it.